src/include/sof/ipc/common.h

```diff
@@ -13,14 +13,32 @@
 #include <sof/list.h>
 #include <sof/schedule/task.h>
 #include <sof/sof.h>
-#include <ipc/header.h>
 #include <user/trace.h>
 #include <stdbool.h>
 #include <stdint.h>
 
 struct dma_sg_elem_array;
 struct ipc_msg;
 
+/* generic IPC header regardless of ABI MAJOR type that is always 4 byte aligned */
+typedef uint32_t ipc_cmd_hdr;
+
+/*
+ * Common IPC logic uses standard types for abstract IPC features. This means all ABI MAJOR
+ * abstraction is done in the IPC layer only and not in the surrounding infrastructure.
+ */
+#if CONFIG_IPC_MAJOR_3
+#include <ipc/header.h>
+#define ipc_get_hdr(x) ((struct sof_ipc_cmd_hdr *)x)
+#elif CONFIG_IPC_MAJOR_4
+#include <ipc4/header.h>
+#define ipc_get_hdr(x) ((union ipc4_message_header *)x)
+#else
+#error "No or invalid IPC MAJOR version selected."
+#endif
+
+#define ipc_put_hdr(x) ((ipc_cmd_hdr *)x)
```

Comment:
get/put are typically associated with refcounts, this doesn't seem to be the case here.

Alternate less ambiguous definitions would be welcome.

Reply:
agree, ```to``` and ```from``` are probably better here.

Reply:
fixed